woocommerce/woocommerce-android
# Update `minSdkVersion` (from `21` to `24`)

### Description


After an internal discussion (see `peaMlT-2f-p2`) and decision (see `peaMlT-2f-p2#comment-107`), this PR upgrades `minSdkVersion` to `24` as less than 0.5% of the users are on the earlier versions of Android.

-----

As part of this `minSdkVersion = 24` upgrade the below changes were also applied in order to fix any additional warnings that this change brought-up:

Warnings Resolution List:

1. [Resolve support annotation usage lint warning for saved state.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/cb55d9f66e1a86176f5effefaa9205f2b36f5cdf)
2. [Resolve obsolete sdk int lint warning for saved state.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/a7e7470f260e941c361786963f88d2a373605345)
3. [Resolve obsolete sdk int lint warnings for all kotlin files.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/19bcae5506649ca0453524645bfd9cd032df3275)
4. [Resolve obsolete sdk int lint warnings for all xml files.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/a4db4cf32f5fe4ac4fb6799d72c3d8278733722f)
5. [Resolve obsolete sdk int lint warning for res values v23.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/f986dacd547abaf73a12d38aca288b3565a48e47)
6. [Resolve obsolete sdk int lint warning for res drawable v23.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/ab1358f639655854d711edaaa23bd6df55e7675a)
7. [Resolve obsolete sdk int lint warning for res drawable v23.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/5ddc47383c8d4358a2406bc6f900c144124ca889)

Warnings Suppression List:

1. [Suppress unused attribute lint warn for data extraction rules.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/7b066fa114ff57794fc2f39c298713a8e28e32c0)
2. [Suppress unused attribute lint warns for order creation.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/d1e649dedf88230227f7de8a2394c0e7835116f8)

Refactor List:

1. [Replace build version code with system version utils checks.](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/4f588b970bc18c65e2ad61d8818e06fad5131dee)

-----

In addition to the above, and as seen from this prerequisite commit [here](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/473d9452e38abbcba25f566a6c4105cc8cceccbf), the `barcode` module related `compile/min/targetSdkVersion` values were, and some of them still, are being declared separately.

I am not sure why this was done like that, if that was an oversight, or done by design, but just like I did with the `minSdkVersion` [change](https://github.com/woocommerce/woocommerce-android/pull/7643/commits/473d9452e38abbcba25f566a6c4105cc8cceccbf), were I explicitly used the globally defined `gradle.ext.minSdkVersion`, I would recommend that the same happens for the other two such properties, especially for `compileSdkVersion 31`, since the globally defined is currently the `compileSdkVersion = 33`.

@0nko I am shameless pinging you here since I saw you adding this new module as part of this 58c599f052afa9a3797e18ae1abf29bad6723260 commit, in the hope that you can maybe help with that. 🙇 

-----

### Testing instructions


- There is nothing much to test here.
- Verifying that all the CI checks are successful should be enough.
- However, if you want to be thorough about reviewing this change, you could:
  - Quickly smoke test the `WooCommerce` app and see if it works as expected.
  - Try and install the `WooCommerce` app on Android 5 (`API 21/22`) and Android 6 (`API 23`) devices and verify that you can't (see [API Levels](https://apilevels.com/)).
